File: structural/typescript/proxy_structured/caching_proxy.ts
export interface Video {
  play(): void;
}

export class RealVideo implements Video {
  private filename: string;
  
  constructor(filename: string) {
    this.filename = filename;
    this.loadVideo();
  }
  
  private loadVideo(): void {
    console.log(`Loading video ${this.filename} from server...`);
  }
  
  play(): void {
    console.log(`Playing video: ${this.filename}`);
  }
}

export class VideoProxy implements Video {
  private filename: string;
  private realVideo?: RealVideo;
  private isLoaded: boolean = false;
  
  constructor(filename: string) {
    this.filename = filename;
  }
  
  play(): void {
    if (!this.isLoaded) {
      console.log('Loading video on first play...');
      this.realVideo = new RealVideo(this.filename);
      this.isLoaded = true;
    } else {
      console.log('Using cached video...');
    }
    this.realVideo!.play();
  }
}

File: structural/dart/proxy_structured/caching_proxy.dart
abstract class Video {
  void play();
}

class RealVideo extends Video {
  final String filename;
  
  RealVideo(this.filename) {
    _loadVideo();
  }
  
  void _loadVideo() {
    print('Loading video $filename from server...');
  }
  
  @override
  void play() {
    print('Playing video: $filename');
  }
}

class VideoProxy extends Video {
  final String filename;
  RealVideo? _realVideo;
  bool _isLoaded = false;
  
  VideoProxy(this.filename);
  
  @override
  void play() {
    if (!_isLoaded) {
      print('Loading video on first play...');
      _realVideo = RealVideo(filename);
      _isLoaded = true;
    } else {
      print('Using cached video...');
    }
    _realVideo!.play();
  }
}
